### PR TITLE
Bug/text size

### LINKED ISSUE #1960: Throw exception if nonexistent keyword arguments are given
There was a typo in the line `glyphs/line.py` example where it said `line_with` instead of `line_width`. Not a big deal in this case, but can lead to problems that are hard to debug. My feeling is that we should raise an error, or at least show a warning when an invalid keyword argument is given.

### LINKED ISSUE #2316: Text glyph font size from ColumnDataSource field not working in 0.9
In 0.9, I can't link the value of the text size of a text glyph to a field in a ColumnDataSource.  This worked in 0.8.1.

Example using 0.8.1: http://jsfiddle.net/e26k8m1c/
Same example using 0.9.0: http://jsfiddle.net/eLsvnd1g/

### PR DESCRIPTION
**Pull request by bryevdv**
issues: fixes #2316 closes #1960 

fortunately was a simpler fix that I had expected.

### CODE REVIEW

**Comment by bryevdv** [comment]
@bokeh/dev this fixes the problem but `text_font_size="8pt"` no longer works because for all "Spec" properties, text is default assumed to be a column name. So, thoughts on how to handle? Could change to
* `text_font_size=value("8pt")`
* `text_font_size=["8pt"]` (maybe we should ditch this "convenience")

Or could put in a special check for strings that look like values (ends with "pt", "em", etc.)

**Comment by bryevdv** [comment]
Also ping @hafen

**Comment by damianavila** [comment]
>Could change all the (now) broken examples to...

Yep, but

>Or could put in a special check for strings that look like values (ends with "pt", "em", etc.) to keep the current syntax "working".

we need this to not break people...

File: bokeh/resources.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def _get_cdn_urls(version=None, minified=True):
     container = dev_container if _DEV_PAT.match(version) else rel_container
 
     if version.endswith(('dev', 'rc')):
-        logger.warn("Getting CDN URL for local dev version will not produce usable URL")
+        logger.debug("Getting CDN URL for local dev version will not produce usable URL")
```
Comment (damianavila) [inline bokeh/resources.py]: Thanks for this... was very annoying...

**Comment by bryevdv** [comment]
OK, so think I think we will have to make a new "Spec" just for text size that can supply a value when any thing that looks like a value comes through. I will try to look at this tonight. What are the units to look for? `"pt"`, `"px"`, `"em"` anything else?

**Comment by damianavila** [comment]
> "pt", "px", "em" anything else?

maybe "rem", "ex"... there are other but not sure how frequently are used...

also in the absolute side, maybe "mm"??

**Comment by bryevdv** [comment]
I wonder if a better test would be "starts with a digit" ?

**Comment by damianavila** [comment]
> I wonder if a better (simpler) test would be "starts with a digit" ?

Yep, probably...

**Comment by birdsarah** [comment]
I don't know the canvas spec. From css-land

xx-small | x-small | small | medium | large | x-large | xx-large | smaller | larger | <length> | <percentage>

Length splits into 

Absolute:
- Inches ( in )
- Centimeters ( cm )
- Millimeters ( mm )
- Points ( pt )
- Picas ( pc )

Relative:
- Em-height ( em )
- Root element em-height ( rem )
- X-height ( ex )
- ZERO width ( ch )
- Pixels ( px )
- Viewport width unit ( vw )
- Viewport height unit ( vh )
- Viewport minimum unit ( vm )

Source: CSS Pocket Reference 4th Edition
-----

I mostly use `em` or `px` in my own css work. I like the idea of us supporting, `small` etc. %age would also be good.

Not sure we have to do all of this now!

**Comment by bryevdv** [comment]
@birdsarah this issue actually isn't about any of that, it is just about the interpretation of bokeh properties that are set to strings. I had hoped we could maintain the simple rule that "all strings are always default interpreted as column names" but it looks like we will need to add at least one exception so that strings that being with a digit (for text size properties) are interpreted as values, not column names.

**Comment by bryevdv** [comment]
As an aside a `TextSizeSpec` would also probably give us the hook to support `%age` and "scaling" text in a simpler way that I had imagined

**Comment by hafen** [comment]
+1 for this!

**Comment by birdsarah** [comment]
@bryevdv - i'm confused. I thought this ticket was saying that we want to keep the spelling text_font_size="8pt" or text_font_size="1.6em" and in order to do that we need a list of possible suffixes.

**Comment by bryevdv** [comment]
Well, your comment about `%age` and "Not sure we have to do all of this now!" implied implementing features that don't exist yet to handle all of those cases that canvas might not handle natively. As you've restated I agree, tho to make things simple were are instead going with the simpler check "starts with a digit" instead of "ends with one of these 37 suffixes"

**Comment by birdsarah** [comment]
@bryevdv sorry about the throw away comments at the end.

Can I make sure I understand the implication of the digit first check though.

If my data looks like this:

x | y | 2colors | 2sizes
------------ | ------------- | ------------ | -------------
1 | 1 | 'white' | '10pt'
2 | 2 | 'pink' | '20pt'

and my glyph

````python
Text(x='x', y='y', text_font_color='2colors', text_font_size='2sizes')
````
Then it won't pick up the 10pt / 20pt it'll try and the text with the size '2sizes.'

**Comment by bryevdv** [comment]
Yes that is correct. Here is my rationale for why that is acceptable (please offer any contrary argument):

* `2sizes` is not a valid python identifier, and is not a valid "dot access" JS identifier either. I think it is less likely (not impossible) that users would try to use that as a column name
* This only (potentially) affects column names for `font_size` properties and nothing else
* It's always possible to explicitly override and stipulate that it is a field: `text_font_size=field('2sizes')`

Given all of that together I am willing to make an exception in the name of both convenience and not breaking existing code. Does everyone agree with this course?

**Comment by birdsarah** [comment]
Here's my thoughts:

1 - This kind of inconsistency is the kind of thing that trips you up when you've either long forgotten about it or first come to a library, and I think it breaks trust ......"what other sneaky little exceptions are there" 

2 - one of my joys of bokeh is the data drivenness of it, its kind of the heart of it, which is why it feels like 
`Text(x='x', y='y', text_font_color='2colors', text_font_size='2sizes')` should be somehow untouchable. `text_font_size='1.6em'` or `'10pt'`,  `'120%'`, `'2cm'` are convenience sugar on top that quickly get converted to `text_font_size=['1.6em']`

3 - when would this happen...the class of problem that springs to mind is not when you're working in a notebook and tab completing your columns, but when you're programmatically generating data & plots. You have a variable name, 'text_size', but it's value is arbitrary, `1345fw57`, and you use that value to both build your data source and your glyph.

4 - it does seem unlikely to me that you would have the following data:

x | y | 2colors | 2pt
------------ | ------------- | ------------ | -------------
1 | 1 | 'white' | '10pt'
2 | 2 | 'pink' | '20pt'

So I'd be ok with just a regex check for valid font-sizes and we don't need to inspect all the column names too.

**Comment by bryevdv** [comment]
Gah, that last commit was supposed to be two commits. Oh well. 

OK well, it sounds like we have something to discuss further. I have put the change in, but we can revert that part easily enough. One thing I will add is I *really* don't like the "single item list" shortcut. It's completely non-discoverable and in no way intuitive. I'd rather like to get rid of it. I did add that `value` and `field` convenience funcs in this PR, I wonder if those would be better to promote. 

That said, what about enabling the "single digit check" for now, but adding a deprecation warning, and also updating all the examples to use `value("12pt")` or `["12pt"]`. That would be easy to do with the new `FontSizeSpec` and I do think have `FontSizeSpec` will be useful in other ways in the future, so it is good to add now.

@damianavila, @bokeh/dev please add your thoughts.

**Comment by bryevdv** [comment]
@birdsarah Also I want to state unequivocally: I very much appreciate your reasoned push-back!

**Comment by birdsarah** [comment]
thanks @bryevdv 

First up, I really like `value()` and `field()` I think that's a big win.

My understanding of this PR is that changing `FontSizeSpec` for a more robust check in `__set__` will be straight forward. And if I care enough about it, I can submit a PR :D

**Comment by bryevdv** [comment]
@birdsarah yes the `__set__` in `FontSizeSpec` should allow for lots of things, for instance if we want to allow scaling font sizes somehow, we could possibly do it via there. Another option would be to make it a subclass of `UnitSpec` so that the size could be specified as number with the the units separately (could keep "jammed together as a string" as an option too, tho).

And we could also add a depreciation that shows how to use `value` right now. Is adding the deprecation something we want to do right now? It would be trivial.

**Comment by birdsarah** [comment]
You're suggesting deprecating `text_font_size='2pt'` and encourage people to use `text_font_size=value('2pt')`?

**Comment by bryevdv** [comment]
I'm saying it is an option. We have to decide what is more important. Iron-clad simple consistency "string value are always interpreted as column names by default", or having the convenience of `text_font_size='2pt'` (can't have both)

**Comment by birdsarah** [comment]
"(can't have both)" ..... grumbles about wanting to have cake and eat it. 

I am certainly in favor of simple consistency.

I do also want the convenience of `text_font_size='2pt'` but with much stricter controls than the single digit check......this brings me all the way back to my first post in this thread where i was proposing that we check for any valid css value.....i now see why this problematic.....

I don't feel great about an answer....deprecation is making my front-end web-dev side sad (it likes  `text_font_size='2pt'` a lot).....this is making me think more broadly about the general feature idea of being able to spec a theme in css/sass and have it meld into bokeh.

If I was forced to pick now, I would deprecate. Hopefully wiser souls will chip in.

**Comment by bryevdv** [comment]
@damianavila @bokeh/core this is ready for review. Comments about the deprecation question above still welcome.

**Comment by almarklein** [comment]
This is probably a bad idea, but what about a prefix in the string to mark it as a value?

    text_font_size='2pt'  # ref to a column
    text_font_size='!2pt'  # value

**Comment by bryevdv** [comment]
OK, I am really done now, modulo the deprecation decision. The glyph function decorator has been upgraded from "completely indecipherable" to "merely convoluted". I want to do more (including an auto-doc directive), but it can wait until #1917

**Comment by damianavila** [comment]
>OK, I am really done now, modulo the deprecation decision. 

In general, this LGTM...

>The glyph function decorator has been upgraded from "completely indecipherable" to "merely a bit convoluted". I want to do more (including an auto-doc directive), but it can wait and be part of #1917

I agree, still convoluted, but it is a progress :wink:

> That said, what about enabling the "single digit check" for now, but adding a deprecation warning, and also updating all the examples to use value("12pt")

:+1:  to this... deprecation warning and use `value`... consistency matters... at the end, we are not writing css :trollface:

**Comment by bryevdv** [comment]
OK I added the deprecation warning. I am going to leave the changing of the examples to a docs PR (that can happen during the freeze)

**Comment by bryevdv** [comment]
TravisCI passing, S3 report looks good, merging.

**Pull request merged**